### ISSUE TITLE
Support new PHP 8.4 syntaxes

### ISSUE THREAD

**Issue by jrfnl** [PHPCSStandards/PHP_CodeSniffer#734]
[PHP 8.4](https://www.php.net/releases/8.4/en.php) introduces new syntaxes for which it should be verified if the Tokenizer needs updates and/or whether any sniffs need updating.

## Action list

Notes:
* While Property Hooks is a single RFC, it introduces multiple new syntaxes. With that in mind, I've broken the RFC up into separately actionable parts.
* _Action items in italics are tentative. Investigation and testing is needed to determine whether any changes are needed and if so, which._
* For action items not in italics, I'm fairly certain those will be needed.
* This list is not set in stone and is subject to change once investigation for supporting a syntax starts.

### To Do

- **[Asymmetric Visibility v2](https://wiki.php.net/rfc/asymmetric-visibility-v2)**
    - [ ] Proposal needed on how to handle this for the tokenizer
    - [ ] Tokenizer changes needed - property declarations
    - [ ] Tokenizer changes needed - constructor property promotion
    - [ ] Updates needed to the predefined token collections in the `Tokens` class
    - [ ] Updates needed to utility functions - `getMemberProperties()`, `getMethodParameters()` (for constructor property promotion)
        These updates should include special handling for the abbreviated form.
    - [ ] _Updates needed to abstract sniffs ?_
        This needs investigation, but I imagine the `AbstractScopeSniff` and the `AbstractVariableSniff` might need updates.
    - [ ] Sniff updates needed
- **[new MyClass()->method() without parentheses](https://wiki.php.net/rfc/new_without_parentheses)**
    - [ ] _Tokenizer changes needed ?_ 
    - [ ] _Updates needed to utility functions ?_
    - [ ] _Sniff updates needed ? (or possibly only sniff test updates?)_
- **[Property hooks](https://wiki.php.net/rfc/property-hooks) - `final` properties**
    - [ ] _Tokenizer changes needed ?_ 
    - [ ] Updates needed to utility functions - `getMemberProperties()`, `getMethodParameters()` (for constructor property promotion)
    - [ ] Sniff updates needed
- **[Property hooks](https://wiki.php.net/rfc/property-hooks) - `abstract` properties**
    - [ ] _Tokenizer changes needed ?_ 
    - [ ] Updates needed to utility functions - `getMemberProperties()`, `getMethodParameters()` (for constructor property promotion)
    - [ ] Sniff updates needed
- **[Property hooks](https://wiki.php.net/rfc/property-hooks) - properties on interfaces**
    - [ ] _Tokenizer changes needed ?_ 
    - [ ] Updates needed to utility functions - `getMemberProperties()`, `getMethodParameters()` (for constructor property promotion)
    - [ ] Sniff updates needed
- **[Property hooks](https://wiki.php.net/rfc/property-hooks) - hook syntax, long form**
    - [ ] Proposal needed on how to handle this for the tokenizer
    - [ ] Tokenizer changes needed
    - [ ] Updates needed to utility functions - `getMemberProperties()`
    - [ ] Sniff updates needed
- **[Property hooks](https://wiki.php.net/rfc/property-hooks) - hook syntax. short (arrow) form**
    - [ ] Proposal needed on how to handle this for the tokenizer
    - [ ] Tokenizer changes needed
    - [ ] Updates needed to utility functions - `getMemberProperties()`
    - [ ] Sniff updates needed
- **[Property hooks](https://wiki.php.net/rfc/property-hooks) - hook syntax. other short form variations**
    - [ ] Proposal needed on how to handle this for the tokenizer
    - [ ] Tokenizer changes needed
    - [ ] Updates needed to utility functions - `getMemberProperties()`
    - [ ] Sniff updates needed
- **[Property hooks](https://wiki.php.net/rfc/property-hooks) - hook syntax for abstract properties/interface properties (no body)**
    - [ ] Proposal needed on how to handle this for the tokenizer
    - [ ] Tokenizer changes needed
    - [ ] _Updates needed to utility functions ?_
    - [ ] Sniff updates needed
- **[Property hooks](https://wiki.php.net/rfc/property-hooks) - hook syntax in constructor property promotion, long form**
    - [ ] Tokenizer changes needed
    - [ ] Updates needed to utility functions - `getMethodParameters()`
    - [ ] Sniff updates needed
- **[Property hooks](https://wiki.php.net/rfc/property-hooks) - hook syntax in constructor property promotion, short form**
    - [ ] Tokenizer changes needed
    - [ ] Updates needed to utility functions - `getMethodParameters()`
    - [ ] Sniff updates needed
- **[Property hooks](https://wiki.php.net/rfc/property-hooks) - availability of "magic" variable within hook functions**
    - [ ] _Updates needed to utility functions ?_ - `getMethodParameters()` ?
    - [ ] _Sniff updates needed ?_

### Done
_Nothing yet._

### No action needed

- [Transform exit() from a language construct into a standard function](https://wiki.php.net/rfc/exit-as-function)
    @jrfnl I've reviewed and ran tests with this change at the time as it isn't a new syntax, but a change to an existing syntax with a high probability to break existing sniffs. Based on the final implementation of the change, no changes to support this in PHPCS were needed.

## Contribution Process

If anyone familiar with the PHP and PHPCS tokenizer wants to start work on any of these, please leave a comment in this ticket to claim an action item.

If you are not familiar with either tokenizer, please spare yourself the pain 😉 

New custom PHPCS tokens may be needed (like for supporting property hooks). If that's the case, please open a separate issue with a detailed proposal on how to handle the syntax, what alternatives have been considered, what the consequences would be and why the proposed option should be accepted. Please open this issue before creating a PR.

If anyone needs it, here are some examples of earlier proposals for supporting PHP syntaxes from before PHP 8.4:
* https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/387
* https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/529#issuecomment-2209337419

## Timeline

For anyone coming to this ticket to see when syntax support for PHP 8.4 syntaxes is expected, please do **NOT** spam this ticket with repeated comments asking for a timeline.
**It will be ready when it's ready and whatever is ready will be included in a release forthwith.**

Seriously, to add syntax support for PHP 8.4, a completely silly amount of work will be needed and people need to sleep once in a while as well (and should get a life), so please be patient.

To quote myself from https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/731#issuecomment-2501515300:

> ... property hooks is the bane of my existence and will be a nightmare to add support for in PHPCS. The tokenizer support alone is going to take months of work, as it is not 1 syntax they have added, but over a dozen new syntaxes, all under one RFC... 
> Only after syntax support is available through the Tokenizer can we start fixing individual sniffs and I expect 60-70% of sniffs in the wider PHPCS field (including external standards) will need updates, or at the very least additional tests with property hooks.
> 
> I spoke up about this at the time (and about other idiocracies in property hooks, it's a massive footgun), but unfortunately to no avail.
> If you want to read my full analysis of the problems with property hooks, have a read through here: https://externals.io/message/123048#123080
> 
> Oh and this comment from me on the Internals mailinglist should also help illuminate why so many sniffs will need updates and how problematic property hooks are: https://externals.io/message/123048#123118
> 
> All in all, please be patient, I expect it will be quite a while before PHPCS will have full support for property hooks.
> 
> [Funding for this would help](https://opencollective.com/php_codesniffer). Finding a way to clone me, would help even more.

**Comment by paulshryock** [comment]
Hello, are property hooks with curly braces supported in PHP_CodeSniffer 4.0.1?

This is getting flagged:

```php
<?php

interface Thing {
  public string $id { get; }
}
```
```bash
4 | ERROR | [x] Curly brace syntax for accessing array elements and string offsets has been deprecated in PHP 7.4 and removed in PHP 8.0. Found: $id { get; }
  |       |     (PHPCompatibility.Syntax.RemovedCurlyBraceArrayAccess.Removed)
```

**Comment by rodrigoprimo** [comment]
Hi @paulshryock,

I'm not a maintainer of this project, so this is not an official response. I'm just going by the issue description and my experience as a contributor to this project.

No, property hooks are not yet supported in PHPCS. As you can see in the action list in the issue description, most items related to property hooks remain unchecked.

The error you are seeing is from the PHPCompatibility standard, not from PHPCS core. I haven't checked, but I believe that since PHPCS doesn't yet understand the property hook syntax, the tokenizer doesn't recognize the curly braces after `$id` as being part of a property hook declaration, and the PHPCompatibility sniff misinterprets them as the deprecated curly brace array access syntax.

**Comment by nlighteneddesign** [comment]
Is there anything a new contributor can do to help with property hook support?

**Comment by jrfnl** [comment]
@nlighteneddesign Well, surfacing test code, especially edge cases which need to be taken into account, would be helpful, but more than anything this needs funding.

I estimate it would take me 2 to 3 months of work to get property hooks properly sorted in all places it needs support and, aside from needing to find the time, I also can't justify spending that amount of time without compensation.
Keep in mind that the last 7 months of 2025 were unpaid too as the budget for the whole year went to getting PHPCS 4.0 released.

**Comment by nlighteneddesign** [comment]
Thank you @jrfnl! I will take a look.

Just a little background, I come from the Drupal community and we are looking to use entity hooks to help with some Fiber related issues we are experiencing.

I saw some interest this morning in assisting with this since it affects the Tokenizer so currently if you use entity hooks you have to disable the whole file from scanning.

Hopefully we can help get this supported!

**Comment by jrfnl** [comment]
Just FYI - in the mean time I've pulled a fix for the [issue reported](https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/734#issuecomment-3887596006) by @paulshryock : https://github.com/PHPCompatibility/PHPCompatibility/pull/2026

Once merged, that should at least prevent that particular false positive (and prevent the fixer from making incorrect fixes).

Not a solution, but still a little bit of a stopgap.

**Comment by paulshryock** [comment]
> that should at least prevent that particular false positive (and prevent the fixer from making incorrect fixes)

Thank you!
